src/main.rs

```diff
@@ -0,0 +1,827 @@
+use std::env;
+use std::format;
+use std::str;
+use std::time::Duration;
+
+use clap::{Command, Arg};
+
+use futures::stream::FuturesUnordered;
+use futures::{StreamExt, TryStreamExt};
+
+use log::{info, error};
+
+use rdkafka::config::{ClientConfig, RDKafkaLogLevel};
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::Consumer;
+use rdkafka::message::{BorrowedMessage, OwnedMessage};
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use rdkafka::Message;
+
+use avro_rs::{from_value};
+use schema_registry_converter::schema_registry_common::SubjectNameStrategy;
+use schema_registry_converter::blocking::avro::{AvroEncoder, AvroDecoder};
+use schema_registry_converter::blocking::schema_registry::{SrSettings};
+
+use chrono::{TimeZone, Utc};
+
+use oxigraph::model::*;
+use oxigraph::store::{Store};
+
+use crate::utils::{setup_logger, log_error_and_return_none};
+
+use crate::vocab::{dcterms, dcat, dcat_mqa, oa};
+
+use crate::schemas::{setup_schemas, DatasetEvent, MQAEvent, MQAEventType};
+use crate::rdf::{
+    parse_turtle, 
+    list_distributions,
+    list_formats,
+    list_media_types,
+    get_dataset_node, 
+    has_property,
+    add_property,
+    convert_term_to_named_or_blank_node_ref,
+    create_metrics_store, 
+    add_quality_measurement,
+    add_five_star_annotation,
+    add_derived_from,
+    is_rdf_format,
+    dump_graph_as_turtle};
+
+use crate::reference_data::{valid_file_type, valid_media_type};
+
+mod utils;
+mod vocab;
+mod schemas;
+mod rdf;
+mod reference_data;
+
+async fn record_borrowed_message_receipt(msg: &BorrowedMessage<'_>) {
+    // Simulate some work that must be done in the same order as messages are
+    // received; i.e., before truly parallel processing can begin.
+    info!("Message received: {}", msg.offset());
+}
+
+async fn record_owned_message_receipt(_msg: &OwnedMessage) {
+    // Like `record_borrowed_message_receipt`, but takes an `OwnedMessage`
+    // instead, as in a real-world use case  an `OwnedMessage` might be more
+    // convenient than a `BorrowedMessage`.
+}
+
+// Read DatasetEvent message of type DATASET_HARVESTED
+fn handle_dataset_event<'a>(msg: OwnedMessage, mut decoder: AvroDecoder) -> Option<MQAEvent> {
+    info!("Handle DatasetEvent on message {}", msg.offset());
+
+    let dataset_event: Option<DatasetEvent> = match decoder.decode(msg.payload()) {
```

Comment:
Tror det blir litt penere om denne returnerer `Result<DatasetEvent, String>`, og at feilmeldingen printes sammen med `error!("Unable to decode dataset event")` i neste `match`. Hva tror du om det, fremfor at den printer error og returerer `None`?

```go
fn parse_dataset_event(msg: OwnedMessage, mut decoder: AvroDecoder) -> Result<DatasetEvent, String> {
    ...
                        "no.fdk.dataset" => match from_value::<DatasetEvent>(&result.value) {
                            Ok(event) => Ok(event),
                            Err(e) => Err(format!("Deserialization failed {}", e)),
                        },
```

Reply:
`Result` return type kan også vurderes på `handle_dataset_event` og `parse_rdf_graph_and_calculate_metrics`.

Reply:
Må andre arg i Result (generics) ikke være av type Error?

Reply:
Ser ikke sånn ut, er en static &'static str i eksempelet: https://doc.rust-lang.org/std/result/